### PR TITLE
[Reviewer: AJH] Update sas-client to use new logging callback

### PR DESCRIPTION
**Pull request by MF1-MS**
Function to convert logs raised within the sas-client to types expected by the new common sas logging callback added. The line number and module are written to the msg arg that is passed to the common callback. 

Logs raised from the SAS-client are displayed just as they were before, which I have verified on an AWS deployment. Due to the mapping between sasclient_log_level_t and Log::logging_level some logs raised in the SAS-Client will appear at a different level than expected, but at the default log level 2 all logs one expects to be raised are.

PR for new logging callback is [here](https://github.com/Metaswitch/cpp-common/pull/755)

### CODE REVIEW

**Comment by rowenasmithlamkin** [comment]
FYI the new client C API does now support analytics messages, so you don't need to remove them from this client.

File: include/sas.h [hunk 1]
```diff
@@ -360,35 +374,93 @@ class SAS
...
+  /// The contents of the supplied marker are unchanged, and the ownership
+  /// remains with the calling code
+  ///
+  /// @param marker
+  ///    The pre-constructed Marker to send
+  /// @param scope
+  ///    The association scope.  One of: NONE, BRANCH, TRACE
+  /// @param reactivate
+  ///    Determines whether we set the association flags byte to 0x01 if true
+  ///    or 0x02 if false (or 0 f the scope was none).
```
Comment (AlexHockey) [inline include/sas.h]: That's not a very useful comment. What does the flag actually mean?

File: include/sas.h [hunk 2]
```diff
@@ -397,8 +469,22 @@ class SAS
...
 
-private:
+  /// Converts the format of a log raised within the SAS-Client to that expected by
+  /// the common log callback, and calls the common log callback with the converted
+  /// arguments.
+  static void sasclient_log_callback(log_level_t level,
+                                     const char *module,
+                                     int line_number,
+                                     const char *fmt,
+                                     ...);
```
Comment (AlexHockey) [inline include/sas.h]: Why does this need to be public.

File: include/sas.h [hunk 3]
```diff
@@ -397,8 +469,22 @@ class SAS
...
+
+  static void _sasclient_log_callback(log_level_t level,
+                                      const char *module,
+                                      int line_number,
+                                      const char *fmt,
+                                      va_list args);
```
Comment (AlexHockey) [inline include/sas.h]: Why is this needed? It's the job of `sasclient_log_callback` to convert between it's  parameters and the parameters that the callback needs so I don;t see why we need to go via this function.

File: source/sas.cpp [hunk 4]
```diff
@@ -756,40 +758,89 @@ SAS::Timestamp SAS::Marker::get_timestamp() const
...
+  char logline[MAX_LOGLINE];
+  int written = 0;
+
+  // Strip the directory from the file location.
+  const char* mod = strrchr(module, '/');
+  module = (mod != NULL) ? mod + 1 : module;
+
+  if (line_number)
+  {
+    written = snprintf(logline, MAX_LOGLINE - 2, "CHECK %s %s:%d: ", log_level[level], module, line_number);
```
Comment (AlexHockey) [inline source/sas.cpp]: Why is there this random CHECK in here?

File: source/sas.cpp [hunk 5]
```diff
@@ -756,40 +758,89 @@ SAS::Timestamp SAS::Marker::get_timestamp() const
...
+    case LOG_LEVEL_VERBOSE:
+      sas_level = SASCLIENT_LOG_DEBUG;
+      break;
+    case LOG_LEVEL_DEBUG:
+      sas_level = SASCLIENT_LOG_DEBUG;
+      break;
+    default:
+      SAS_LOG_WARNING("Unknown SAS log level %d, treating as error level", level);
+      sas_level = SASCLIENT_LOG_ERROR;
+    }
```
Comment (AlexHockey) [inline source/sas.cpp]: Do you need to do this conversion? Why not eliminate the internal `LOG_LEVEL_*` variables entirely and just use the `SASCLIENT_LOG_*` constants throughout?

File: source/sas.cpp (hunk 4 above)
Comment (AlexHockey) [reply source/sas.cpp]: The log level gets passed through to the user on the log callback, so you don't need to render it to text here.

File: source/sas.cpp [hunk 6]
```diff
@@ -756,40 +758,89 @@ SAS::Timestamp SAS::Marker::get_timestamp() const
...
+{
+  const char* log_level[] = {"Error", "Warning", "Status", "Info", "Verbose", "Debug"};
+  char logline[MAX_LOGLINE];
+  int written = 0;
+
+  // Strip the directory from the file location.
+  const char* mod = strrchr(module, '/');
+  module = (mod != NULL) ? mod + 1 : module;
+
+  if (line_number)
```
Comment (AlexHockey) [inline source/sas.cpp]: Do we need this test? When do you ever not know the line number?

File: source/sas.cpp [hunk 7]
```diff
@@ -756,40 +758,89 @@ SAS::Timestamp SAS::Marker::get_timestamp() const
...
+  // written if their second argument was large enough, so we need to
+  // reduce the size of written to compensate if it is too large.
+  written = std::min(written, MAX_LOGLINE - 2);
 
-  va_end(args);
-}
+  int bytes_available = MAX_LOGLINE - written - 2;
+  written += vsnprintf(logline + written, bytes_available, fmt, args);
 
+  if (written > (MAX_LOGLINE - 2))
```
Comment (AlexHockey) [inline source/sas.cpp]: What is it always `MAX_LOGLINE - 2` Why the `-2`? Also why do you do `std::min(written, MAX_LOGLINE - 2);` above, but an if test here? Please can you be consistent?

Comment (MF1-MS) [reply source/sas.cpp]: The -2 is there to leave space for the new line and null termination that gets added just before the array is handed over to the logger in log.cpp. I'll add a comment.

File: source/sas.cpp (hunk 6 above)
Comment (MF1-MS) [reply source/sas.cpp]: Likely not in the scope here, I'll remove.

File: source/sas.cpp (hunk 5 above)
Comment (MF1-MS) [reply source/sas.cpp]: Because when this gets passed to the CPP-common log callback we need to check whether or not to write the log against the global logging level.

File: include/sas.h (hunk 2 above)
Comment (MF1-MS) [reply include/sas.h]: To allow the SAS_LOG macro in sas_internal.h access to it.

Comment (AlexHockey) [reply include/sas.h]: Macros don't care about access permissions. They are expanded before the C++ compiler even gets run.

File: source/sas.cpp (hunk 5 above)
Comment (AlexHockey) [reply source/sas.cpp]: What I'm saying is with this PR we have three logging levels (using error logs as an example):

* `LOG::ERROR_LEVEL` defined in log.h
* `SASCLIENT_LOG_ERROR` which is exposed on the sas-client interface. 
* `SAS::LOG_LEVEL_ERROR` which is used in the `SAS_LOG_ERROR` macro. 

Why not get rid of the last set, and have `SAS_LOG_ERROR` just use the values defined on the interface?

File: source/sas.cpp (hunk 7 above)
Comment (AlexHockey) [reply source/sas.cpp]: Ah OK. NULL termination makes sense. But I don't think this function should be adding line breaks.

File: include/sas.h (hunk 2 above)
Comment (MF1-MS) [reply include/sas.h]: Hmmm my compiler is throwing wobbly's I don't understand for now then, it explicitly complains at me if this is private, and is only called by that macro.

File: source/sas.cpp (hunk 5 above)
Comment (MF1-MS) [reply source/sas.cpp]: I see, I agree.

File: source/sas.cpp (hunk 7 above)
Comment (MF1-MS) [reply source/sas.cpp]: It's not adding line breaks, but it's saving the space for when the logging framework needs to.

File: include/sas.h [hunk 8]
```diff
@@ -360,35 +374,93 @@ class SAS
...
+  ///
+  /// @param  system_name
+  ///     The unique name for the system, eg: hostname
+  /// @param  system_type
+  ///     The type of this system
+  ///
+  /// @param  resource_identifier
+  ///     The version of the resource bundle
+  /// @param  sas_address
+  ///     Takes a single ipv4 address
```
[the remark this answers is not in the document]
Comment (MF1-MS) [reply include/sas.h]: I removed no code to that effect, and the supplied sas_address does indeed get compared to a host_map somewhere. Will amend.

File: source/sas.cpp (hunk 7 above)
Comment (AlexHockey) [reply source/sas.cpp]: But the logging framework is using a different buffer to sprintf into, right?

Comment (MF1-MS) [reply source/sas.cpp]: Hmmmm if we give something too large to the logging framework it'll just truncate it as expected, which seems fine. But I have to declare an array of *some* size to sprintf into, so is it not sensible to make it the "correct" size?

Comment (AlexHockey) [reply source/sas.cpp]: The logging framework can and will prepend some stuff onto the front of the log, so this function has no idea what the "right" size is.

I think my ultimate feedback is that this fucntion has too much responsibility. It should just be converting the parameters it receives into a format that allows it to call back into the user of the library. So all it needs to do is sprintf the parameters into a buffer and call `_log_callback`. Let the log library handle printing the level, newlines and whatever else it needs to. When viewed like this, it doesn't make sense for this function to second guess what the log framework is going to do.
